File: script/tools/read_device.c
#include <stdio.h>
#include <stdlib.h>
#include <stdint.h>
#include <unistd.h>
#include <fcntl.h>
#include <sys/ioctl.h>
#include <linux/i2c-dev.h>
#include <linux/i2c.h>
#include <string.h>
#include <i2c/smbus.h>

// Device file
#define I2C_DEV "/dev/i2c-1"
#define I2C_ADDR 0x17

// Define the structure
typedef struct __attribute__((packed)) {
    uint8_t WHO_AM_I; // Fixed to 0xA6
    uint8_t version; // Version (Read only)
    uint32_t uuid0; // Universal unique ID
    uint32_t uuid1;
    uint32_t uuid2;

    uint16_t output_voltage; // Output voltage(5V)
    uint16_t input_voltage; // Input voltage
    uint16_t battery_voltage; // Battery Voltage
    uint16_t mcu_voltage; // MCU voltage
    uint16_t output_current; // Output current (5V)
    uint16_t input_current; // Input current
    int16_t battery_current; // Battery current (Negative value is for discharging, positive value is for charging)
    int8_t temperature; // Temperature of MCU

    // cr - control registers
    union {
        struct {
            uint8_t auto_start_mode : 1; // auto start mode
            uint8_t rsv1 : 7; // Reserved
        };
        uint8_t cr1; // Use to access full byte.
    };
    uint8_t cr2; // Reserved Control Register 2

    // sr - status registers
    union {
        struct {
            uint8_t sw_status : 1; // 5V OUTPUT status (0=Off,1=On)
            uint8_t fast : 1; // Fast charging status (0=slow charging/no external power, 1=12V fast charging)
            uint8_t charge : 1; // Charge/discharge status (0=charging, 1=discharging)
            uint8_t input_low : 1; // Input voltage low (0=normal, 1=low)
            uint8_t output_low : 1; // 5V output low (0=normal, 1=low)
            uint8_t battery_low : 1; // Battery voltage low (0=normal, 1=low)
            uint8_t adc_mismatch : 1; // ADC mismatch (0=normal, 1=INA219 vs MCU sampling difference exceeded)
            uint8_t battery_fail : 1; // Battery failure (0=normal, 1=failure)
        };
        uint8_t sr1; // Used to access the entire byte
    };
    union {
        struct {
            uint8_t python_exec_toobig : 1; // Python code is too big
            uint8_t rsv2 : 7; // Reserved
        };
        uint8_t sr2; // Used to access the entire byte
    };

    uint16_t battery_protection_voltage; // Used to setting the battery protection voltage
    uint16_t shutdown_countdown; // Used to setting shutdown countdown in seconds
    uint16_t auto_start_voltage; // Automatic power-on voltage threshold setting for battery
    uint16_t rsv; // Reserved (Read/Write)
    uint16_t ota_request; // OTA request (Write-only)
    uint64_t runtime; // Accumulated operating time
    uint16_t charge_detect_interval_s; // Charging chip trigger interval (seconds, readable and writable).
    union {
        struct {
            uint8_t i2c_ack : 1; // Light up when I2C communication occurs
            uint8_t bat_charge : 1; // Light up when the battery is charging
            uint8_t bat_discharge : 1; // Light up when the battery is discharging
            uint8_t fault_report : 1; // Light up when there is a fault
            uint8_t ok_report : 1; // Light up during normal operation
            uint8_t rsv3 : 3; // Reserved
        };
        uint8_t led_ctl; // Used to access the entire byte
    };
} DeviceStatus;

// Debug print function
void debug_print(DeviceStatus* status) {
    printf("🌟 Read Device Status Information🌟\n");

    // 打印每个字段
    printf("WHO_AM_I:             0x%02X\n", status->WHO_AM_I);
    printf("Version:              0x%02X\n", status->version);
    printf("UUID:                 0x%08X 0x%08X 0x%08X\n", status->uuid0, status->uuid1, status->uuid2);
    printf("Output Voltage:       %u mV\n", status->output_voltage);
    printf("Input Voltage:        %u mV\n", status->input_voltage);
    printf("Battery Voltage:      %u mV\n", status->battery_voltage);
    printf("MCU Voltage:          %u mV\n", status->mcu_voltage);
    printf("Output Current:       %u mA\n", status->output_current);
    printf("Input Current:        %u mA\n", status->input_current);
    printf("Battery Current:      %d mA\n", status->battery_current);
    printf("Temperature:          %d °C\n", status->temperature);

    printf("Control Register 1 (cr1):       0x%02X\n", status->cr1);
    printf("    Power-on auto-start mode :  %d\n", status->auto_start_mode);

    // printf("Control Register 2 (cr2):       0x%02X\n", status->cr2);

    printf("Status Register 1 (sr1):       0x%02X\n", status->sr1);
    printf("    5V Output status     : %s\n", status->sw_status ? "ON" : "OFF");
    printf("    Fast Charge Mode     : %s\n", status->fast ? "Fast Charge" : "Slow Charge/No External Power");
    printf("    Charging Status      : %s\n", status->charge ? "Discharging" : "Charging");
    printf("    Input Voltage Low    : %s\n", status->input_low ? "LOW" : "Normal");
    printf("    Output Voltage LOW   : %s\n", status->output_low ? "LOW" : "Normal");
    printf("    Battery Voltage LOW  : %s\n", status->battery_low ? "LOW" : "Normal");
    printf("    ADC error            : %s\n", status->adc_mismatch ? "YES" : "Normal");
    printf("    Battery failure      : %s\n", status->battery_fail ? "YES" : "NO");

    //printf("Status Register 2 (sr2):       0x%02X\n", status->sr2);
    //printf("    Python Code is too big : %s\n", status->python_exec_toobig ? "YES" : "NO");

    printf("Battery protection voltage : %u mV\n", status->battery_protection_voltage);
    printf("Shutdown Countdown :  %u s\n", status->shutdown_countdown);
    printf("Power-on auto-start battery voltage threshold : %u mV\n", status->auto_start_voltage);
    printf("Reserved Register :    0x%04X\n", status->rsv);
    printf("Request OTA(Over-The-Air update) :   0x%04X\n", status->ota_request);
    printf("Accumulated operating time :  %llu microseconds\n", status->runtime);
    printf("Charging chip trigger interval :     %u seconds\n", status->charge_detect_interval_s);
    printf("LED Control:               0x%02X\n", status->led_ctl);

    printf("LED Status Parsing\n");
    // LED control parsing
    printf("    Light up when I2C communication occurs   : %d\n", status->i2c_ack);
    printf("    Light up when the battery is charging    : %d\n", status->bat_charge);
    printf("    Light up when the battery is discharging : %d\n", status->bat_discharge);
    printf("    Light up when there is a fault           : %d\n", status->fault_report);
    printf("    Light up during normal operation         : %d\n", status->ok_report);
}

int main() {
    int fd;
    DeviceStatus status; // structure of device status storage

    // Open I2C device file
    if ((fd = open(I2C_DEV, O_RDWR)) < 0) {
        perror("Could not open I2C device. ");
        return -1;
    }

    // Setting I2C slave address
    if (ioctl(fd, I2C_SLAVE, I2C_ADDR) < 0) {
        perror("Setting I2C slave failure.");
        close(fd);
        return -1;
    }

    printf("🔄 Start reading device status...\n");

    // Use SMBus read the data multiple times
    for (int offset = 0; offset < sizeof(DeviceStatus); offset += 32) {
        int len = 32;
        if (offset + len > sizeof(DeviceStatus)) {
            len = sizeof(DeviceStatus) - offset; // last part of data
        }

        // Use SMBus to read data
        union i2c_smbus_data data;
        struct i2c_smbus_ioctl_data args = {
            .read_write = I2C_SMBUS_READ,
            .command = offset,
            .size = I2C_SMBUS_I2C_BLOCK_DATA,
            .data = &data
        };

        // Set the number of bytes to read
        data.block[0] = len;

        // Execute SMBus to read data
        if (ioctl(fd, I2C_SMBUS, &args) < 0) {
            perror("SMBus read failed.");
            close(fd);
            return -1;
        }

        // Store reading data to the data structure.
        memcpy((uint8_t*)&status + offset, data.block + 1, len);
        printf("📝 Read from: 0x%02X data size: %d byte\n", offset, len);
    }

    // Print device status
    debug_print(&status);

    // Close I2C device
    close(fd);
    return 0;
}

File: script/tools/python_demo/read_device.py
import smbus2
import struct
from typing import NamedTuple
from dataclasses import dataclass

# Device configuration
I2C_DEV = 3  # /dev/i2c-3
I2C_ADDR = 0x17


@dataclass
class DeviceStatus:
    # Define the fields corresponding to the C structure
    WHO_AM_I: int  # Fixed to 0xA6
    version: int  # Version number (read-only)
    uuid0: int  # Unique ID
    uuid1: int
    uuid2: int

    output_voltage: int  # Output voltage (5V)
    input_voltage: int  # Input voltage
    battery_voltage: int  # Battery voltage
    mcu_voltage: int  # MCU voltage
    output_current: int  # Output current (5V)
    input_current: int  # Input current
    battery_current: int  # Battery current (negative value indicates consumption, positive value indicates charging)
    temperature: int  # Temperature

    cr1: int  # Control register 1
    cr2: int  # Control register 2

    sr1: int  # Status register 1
    sr2: int  # Status register 2

    battery_protection_voltage: int  # Battery protection voltage
    shutdown_countdown: int  # Shutdown countdown
    auto_start_voltage: int  # Battery voltage threshold for auto-start on incoming calls
    pika_output_len: int  # Python output buffer size
    ota_request: int  # Request OTA
    runtime: int  # Cumulative runtime time
    charge_detect_interval_s: int  # Charging chip trigger interval
    led_ctl: int  # LED control


def debug_print(status: DeviceStatus):
    """Print device status information"""
    print("🌟 Read device status information 🌟\n")

    # Print each field
    print(f"WHO_AM_I: 0x{status.WHO_AM_I:02X}")
    print(f"Version number: 0x{status.version:02X}")
    print(f"UUID: 0x{status.uuid0:08X} 0x{status.uuid1:08X} 0x{status.uuid2:08X}")
    print(f"Output voltage: {status.output_voltage} mV")
    print(f"Input voltage: {status.input_voltage} mV")
    print(f"Battery voltage: {status.battery_voltage} mV")
    print(f"MCU voltage: {status.mcu_voltage} mV")
    print(f"Output current: {status.output_current} mA")
    print(f"Input current: {status.input_current} mA")
    print(f"Battery current: {status.battery_current} mA")
    print(f"Temperature: {status.temperature} °C")

    print(f"Control register 1 (cr1): 0x{status.cr1:02X}")
    print(f"Incoming call auto-start mode: {(status.cr1 >> 0) & 1}")
    print(f"Loading Python code: {(status.cr1 >> 1) & 1}")
    print(f"Running Python code: {(status.cr1 >> 2) & 1}")
    print(f"Reading Python output log: {(status.cr1 >> 3) & 1}")
    print(f"Reserved for future use: {(status.cr1 >> 4) & 0xF}")

    print(f"Control register 2 (cr2): 0x{status.cr2:02X}")

    print(f"Status register 1 (sr1): 0x{status.sr1:02X}")
    print(f"5V output status : {'On' if (status.sr1 >> 0) & 1 else 'Off'}")
    print(
        f" Fast charge mode: {'Slow charge' if (status.sr1 >> 1) & 1 else 'Fast charge'}"
    )
    print(f" Charging status: {'Discharging' if (status.sr1 >> 2) & 1 else 'Charging'}")
    print(f" Input voltage low: {'Low' if (status.sr1 >> 3) & 1 else 'Normal'}")
    print(f" Output voltage low: {'Low' if (status.sr1 >> 4) & 1 else 'Normal'}")
    print(f" Battery voltage low: {'Low' if (status.sr1 >> 5) & 1 else 'Normal'}")
    print(f" ADC error: {'Yes' if (status.sr1 >> 6) & 1 else 'Normal'}")
    print(f"Battery failure: {'Yes' if (status.sr1 >> 7) & 1 else 'No'}")

    print(f"Status register 2 (sr2): 0x{status.sr2:02X}")
    print(f"Python code too large: {'Yes' if (status.sr2 >> 0) & 1 else 'No'}")

    print(f"Battery protection voltage: {status.battery_protection_voltage} mV")
    print(f"Shutdown countdown: {status.shutdown_countdown} s")
    print(
        f"Incoming call auto-start battery voltage threshold: {status.auto_start_voltage} mV"
    )
    print(f"Python output buffer size: {status.pika_output_len} B")
    print(f"OTA request: 0x{status.ota_request:04X}")
    print(f"Cumulative running time: {status.runtime} milliseconds")
    print(f"Charging chip trigger interval: {status.charge_detect_interval_s} seconds")
    print(f"LED control: 0x{status.led_ctl:02X}")

    print("LED status analysis")
    # LED control analysis
    print(f" Turns on when I2C communication occurs: {(status.led_ctl >> 0) & 1}")
    print(f" Turns on when the battery is charging: {(status.led_ctl >> 1) & 1}")
    print(f" Turns on when the battery is discharging: {(status.led_ctl >> 2) & 1}")
    print(f" Turns on when a fault occurs: {(status.led_ctl >> 3) & 1}")
    print(
        f" Turns on when the device is operating normally: {(status.led_ctl >> 4) & 1}"
    )


def read_device_status() -> DeviceStatus:
    """Read device status"""
    # Define structure format (little endian, consistent with C structure)
    # B: uint8_t, H: uint16_t, I: uint32_t, Q: uint64_t, h: int16_t, b: int8_t
    fmt = "<"  # Little endian
    fmt += "B"  # WHO_AM_I
    fmt += "B"  # version
    fmt += "III"  # uuid0-2
    fmt += "HHHHHH"  # output_voltage, input_voltage, battery_voltage, mcu_voltage, output_current, input_current
    fmt += "h"  # battery_current
    fmt += "b"  # temperature
    fmt += "BB"  # cr1, cr2
    fmt += "BB"  # sr1, sr2
    fmt += "HHHHH"  # battery_protection_voltage, shutdown_countdown, auto_start_voltage, pika_output_len, ota_request
    fmt += "Q"  # runtime
    fmt += "H"  # charge_detect_interval_s
    fmt += "B"  # led_ctl

    # Calculate structure size
    struct_size = struct.calcsize(fmt)
    print(f"Structure size: {struct_size} bytes")

    # Initialize the I2C bus
    bus = smbus2.SMBus(I2C_DEV)

    # Read data in multiple passes, up to 32 bytes each
    data = bytearray()
    for offset in range(0, struct_size, 32):
        # Calculate the length of this read
        read_len = min(32, struct_size - offset)

        try:
            # Read block data using SMBus
            block_data = bus.read_i2c_block_data(I2C_ADDR, offset, read_len)

            # Add to data buffer
            data.extend(block_data)
            print(f"📝 Read data from address 0x{offset:02X}: {len(block_data)} bytes")
        except Exception as e:
            print(f"❌ Error reading address 0x{offset:02X}: {e}")
            # Pad with zeros to ensure data length is correct
            data.extend(bytes([0] * read_len))

    # Check data length is correct
    if len(data) < struct_size:
        print(
            f"⚠️ Insufficient data, expected {struct_size} bytes, received {len(data)} bytes"
        )
        # Pad with zeros
        data.extend(bytes([0] * (struct_size - len(data))))

    # Unpack data into a structure
    try:
        unpacked = struct.unpack(fmt, data[:struct_size])
        return DeviceStatus(*unpacked)
    except Exception as e:
        print(f"❌ Error unpacking data: {e}")
        print(f"Data length: {len(data)} bytes")
        print(f"Format string: {fmt} (expected {struct_size} bytes)")
        raise


if __name__ == "__main__":
    print("🔄 Start reading device data...")
    try:
        status = read_device_status()
        debug_print(status)
    except Exception as e:
        print(f"❌ Error reading device data: {e}")
